lib/screens/v2/disruption_diagram/builder.ex

```diff
@@ -0,0 +1,967 @@
+    branch =
+      branch_letter
+      |> String.downcase()
+      |> String.to_existing_atom()
+
+    {:ok, route_id, sequence, branch}
+  end
+
+  # A single Red Line branch
+  defp do_get_data([{"Red", sequence}], _) do
+    branch = if "place-asmnl" in sequence, do: :ashmont, else: :braintree
+
+    {:ok, "Red", sequence, branch}
+  end
+
+  # A single non-branching route
+  defp do_get_data([{route_id, sequence}], _) do
+    {:ok, route_id, sequence, :trunk}
+  end
+
+  # 2+ routes
+  defp do_get_data(matches, informed_route_id) do
+    cond do
+      Enum.all?(matches, &match?({"Green-" <> _, _}, &1)) ->
+        # Green Line trunk
+        {:ok, "Green", gl_trunk_stop_sequence(), :trunk}
+
+      Enum.all?(matches, &match?({"Red", _}, &1)) ->
+        # Red Line trunk
+        {:ok, "Red", rl_trunk_stop_sequence(), :trunk}
+
+      # The remaining cases are for when 2+ lines contain the stop(s). We defer to informed route.
+      # Only core stops are served by more than one line, so we'll use the trunk sequences for GL/RL.
+      String.starts_with?(informed_route_id, "Green") ->
+        # Green Line trunk, probably at North Station, Haymarket, Government Center, or Park Street
+        {:ok, "Green", gl_trunk_stop_sequence(), :trunk}
+
+      informed_route_id == "Red" ->
+        # Red Line trunk, probably at Park Street or Downtown Crossing
+        {:ok, "Red", rl_trunk_stop_sequence(), :trunk}
+
+      true ->
+        # Orange Line, probably at North Station, Haymarket, State, or Downtown Crossing
+        # or Blue Line, probably at Government Center or State
+        {:ok, informed_route_id, Stop.get_route_stop_sequence(informed_route_id), :trunk}
+    end
+  end
+
+  defp gl_trunk_stop_sequence do
+    Enum.map(Stop.gl_trunk_stops(), fn {stop_id, _labels} -> stop_id end)
+  end
+
+  defp rl_trunk_stop_sequence do
+    Enum.map(Stop.rl_trunk_stops(), fn {stop_id, _labels} -> stop_id end)
+  end
+
+  # Adjusts the left and right ends of the sequence before we split them off into `left_end` and `right_end`.
+  # - Mark terminal stops as such
+  # - For branching ends of trunk sequences (JFK, Lechmere, Kenmore), add `ArrowSlot`s with labels for those branches.
+  defp adjust_ends(sequence, line, branch)
+
+  defp adjust_ends(sequence, :green, :trunk) do
+    # The Green Line trunk (Lechmere to Kenmore) has branches at both ends.
+    sequence
+    |> Vector.prepend(%ArrowSlot{label_id: "place-mdftf+place-unsqu"})
+    |> Vector.append(%ArrowSlot{label_id: "western_branches"})
+  end
+
+  defp adjust_ends(sequence, :red, :trunk) do
+    # The Red Line trunk (Alewife to JFK) has a terminal at Alewife and branches past JFK.
+    sequence
+    |> Vector.update_at!(0, &%{&1 | terminal?: true})
+    |> Vector.append(%ArrowSlot{label_id: "place-asmnl+place-brntn"})
+  end
+
+  defp adjust_ends(sequence, _line, _branch) do
+    # All other stop sequences have terminals at both ends.
```

Comment:
> Is it true that when get_builder_data returns a branch, that means both ends of the sequence are terminals?

Yes, at least at the start.

This function serves to put in arrow slots where we are 100% sure it's a GL/RL trunk alert with the home stop also on the trunk. In those cases, we never want the diagram to go beyond a branch point (Lechmere, Kenmore, JFK), because we wouldn't know which branch to choose, so the relevant ends have to use arrows.

In all other cases, we simply mark the first/last stops as terminals. In subsequent logic they may yet be replaced with arrow slots.